File: app/src/main/java/com/example/clearentwrapperdemo/feature/ReaderInformationFragment.java
package com.example.clearentwrapperdemo.feature;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;

import com.clearent.idtech.android.wrapper.ClearentWrapper;
import com.clearent.idtech.android.wrapper.listener.ReaderStatusListener;
import com.clearent.idtech.android.wrapper.model.ReaderState;
import com.clearent.idtech.android.wrapper.model.ReaderStatus;
import com.example.clearentwrapperdemo.R;
import com.example.clearentwrapperdemo.databinding.FragmentReaderInformationBinding;

public class ReaderInformationFragment extends Fragment implements ReaderStatusListener {

    private static final String noReaderConnectedMessage = "No reader is connected";
    private static final String readerConnectedMessage = "Connected";
    private static final String readerNotConnectedMessage = "Disconnected";

    private FragmentReaderInformationBinding binding;

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        binding = FragmentReaderInformationBinding.inflate(inflater, container, false);
        return binding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        ClearentWrapper.INSTANCE.addReaderStatusListener(this);
    }

    private void renderReaderStatus(ReaderState readerState) {
        if (readerState instanceof ReaderState.NoReader) {
            binding.readerName.setText(noReaderConnectedMessage);
            binding.readerConnectionStatus.setVisibility(TextView.GONE);
            binding.readerSignalAndBattery.setVisibility(TextView.GONE);
        } else if (readerState instanceof ReaderState.ReaderUnavailable) {
            binding.readerName.setText(((ReaderState.ReaderUnavailable) readerState).getReader().getDisplayName());
            binding.readerConnectionStatus.setVisibility(TextView.VISIBLE);
            binding.readerConnectionStatus.setText(readerNotConnectedMessage);
            binding.readerSignalAndBattery.setVisibility(TextView.GONE);
        } else if (readerState instanceof ReaderState.ReaderPaired) {
            binding.readerName.setText(((ReaderState.ReaderPaired) readerState).getReader().getDisplayName());
            binding.readerConnectionStatus.setVisibility(TextView.VISIBLE);
            binding.readerConnectionStatus.setText(readerConnectedMessage);
            binding.readerSignalAndBattery.setVisibility(TextView.VISIBLE);
            binding.readerSignalAndBattery.setText(
                    getString(
                            R.string.reader_signal_and_battery_format,
                            ((ReaderState.ReaderPaired) readerState).getBattery().getBatteryLevel(),
                            String.valueOf(((ReaderState.ReaderPaired) readerState).getSignal().getSignalLevel())
                    )
            );
        }
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        binding = null;
        ClearentWrapper.INSTANCE.removeReaderStatusListener(this);
    }

    @Override
    public void onReaderStatusUpdate(@Nullable ReaderStatus readerStatus) {
        renderReaderStatus(ReaderState.Companion.fromReaderStatus(readerStatus));
    }
}

File: app/src/main/java/com/example/clearentwrapperdemo/feature/ReaderInformationFragment.kt
package com.example.clearentwrapperdemo.feature

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.fragment.app.Fragment
import androidx.lifecycle.lifecycleScope
import com.clearent.idtech.android.wrapper.SDKWrapper
import com.clearent.idtech.android.wrapper.listener.ReaderStatusListener
import com.clearent.idtech.android.wrapper.model.ReaderState
import com.clearent.idtech.android.wrapper.model.ReaderStatus
import com.example.clearentwrapperdemo.R
import com.example.clearentwrapperdemo.databinding.FragmentReaderInformationBinding
import kotlinx.coroutines.launch

class ReaderInformationFragment : Fragment(), ReaderStatusListener {

    companion object {
        private const val noReaderConnectedMessage = "No reader is connected"
        private const val readerConnectedMessage = "Connected"
        private const val readerNotConnectedMessage = "Disconnected"
    }

    private var _binding: FragmentReaderInformationBinding? = null
    private val binding get() = _binding!!

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        _binding = FragmentReaderInformationBinding.inflate(inflater, container, false)

        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        SDKWrapper.addReaderStatusListener(this)
    }

    private fun renderReaderStatus(readerState: ReaderState) = lifecycleScope.launch {
        binding.apply {
            when (readerState) {
                ReaderState.NoReader -> {
                    readerName.text = noReaderConnectedMessage
                    readerConnectionStatus.visibility = TextView.GONE
                    readerSignalAndBattery.visibility = TextView.GONE
                }
                is ReaderState.ReaderUnavailable -> {
                    readerName.text = readerState.reader.displayName
                    readerConnectionStatus.visibility = TextView.VISIBLE
                    readerConnectionStatus.text = readerNotConnectedMessage
                    readerSignalAndBattery.visibility = TextView.GONE
                }
                is ReaderState.ReaderPaired -> {
                    readerName.text = readerState.reader.displayName
                    readerConnectionStatus.visibility = TextView.VISIBLE
                    readerConnectionStatus.text = readerConnectedMessage
                    readerSignalAndBattery.visibility = TextView.VISIBLE
                    readerSignalAndBattery.text = getString(
                        R.string.reader_signal_and_battery_format,
                        readerState.battery.batteryLevel,
                        readerState.signal.signalLevel.toString()
                    )
                }
            }
        }
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
        SDKWrapper.removeReaderStatusListener(this)
    }

    override fun onReaderStatusUpdate(readerStatus: ReaderStatus?) {
        renderReaderStatus(ReaderState.fromReaderStatus(readerStatus))
    }
}